### DISCUSSION TITLE
2024 post-season TODOS

### DISCUSSION THREAD

**Discussion by lauckhart** [IronRiders/software-team#1] [Ideas]
Here's some high-level thoughts for next steps for the software team to consider now that we've elected next year's leads.

1. Create (or update if we have one?) a "4180 software engineering 101 - getting started" guide for new devs
    - Describe where to go for help
    - Document how to set up computers for 4180 development (game tools, IDE, git)
    - Standardize on and document online resources for learning the basics of Java
    - Point to online resources for learning the basics of wpilib, vendor and other 3rd party libraries we use
    - Give an overview of key 4180 GitHub repositories and codebase
    - Introduction to basic tasks (build, deploy, debug, drive, simulate, etc.)

2. Discuss what kinds of learning activities we want to do for the rest of the year and/or over the summer. I'm available & happy to help out here in whatever capacity you guys would like

3. Consider establishing a formal code review process
    - Require commits to "main" branch to go through a PR?
    - Would be great if PRs cannot be merged until they've been reviewed and approved by at least one other engineer
    - This serves both as a chance for developers to get feedback and for the reviewer to learn from the original author
    - This is a lightweight process that's built into GitHub. It's pretty standard for open-source and corporate teams of our size

4. Identify individual team members that are interested in specializing and/or taking responsibility for different areas of functionality. Could divide up in various ways but here's some suggestions for high-level categories:
    - Drive & control
    - Manipulation
    - Vision ("pose estimation" via april tags, object detection)
    - Automatous features (autos and driver-assist)
    - Lights

5. Make sure we have enough test platforms so that everybody who's interested can learn on actual hardware
    - Make sure there's one working test board for gaining experience with RoboRIO. Keep it wired with some basic components (e.g. LEDs and at least one motor)
    - Always have at least one drive base available so engineers can gain familiarity with coding swerve drive
    - Get some small robot kits. I've acquired a Romi with a manipulator arm and a couple of cameras. We could get more Romis or the cheaper XRP for the club. These aren't swerve drive but should still be good teaching tools

6. Select an IDE to standardize on going forward
    - Personally I'd use VS code because it's the standard for FRC and entirely free. Just makes everything easier if you're reading FRC docs on how to do something... They reference VS Code a lot, so if using e.g. IntelliJ not only do you need to find the docs but you need to figure out how to do the same thing in IntelliJ
    - From a features and functionality perspective (I've used both IntelliJ and VS Code extensively)... I can think of small things that I appreciate about both but nothing that gives me a strong preference for one vs. the other

7. Consider creating a library we can use between seasons. This would allow for code reuse so for example if our drive base remains essentially the same we don't need to redo the swerve implementation every time we code for a new bot

8. Discuss basic task management (e.g. where to put lists like this one)... GitHub project? Issues? Discussions? Stick w/ Discord?

**Comment by Saturncorgi** [comment]
Regarding number 8 it would be nice to not use discord for things needed while at school as it’s inaccessible on school WiFi.
Additionally it would be nice to find a way to not require software members to own their own laptops. For some reason GitHub is only blocked on sps devices but not the WiFi. School laptops also can't run unapproved apps and that list does not as far as I know include any IDEs. I think that could scare off members unwilling or unable to afford their own device.

**Comment by ryanelbert5125gmail** [comment]
Meeting Minutes 5/8/24

Planning on starting with coding Cubert
Create a repository with tools we would use every year (mainly for swerve)
Next year have people do a pull request before merging into main
Using the Git Hub Project for Project Management
Using VS code as a standard IDE
Using GitHub Discussions so we can make things permanent and being able to access it on school WIFI
Having the 4180 software engineering 101 be on the git hub but having the copies on the drive (DATE IN NAME to prevent old copies)

**Reply by ryanelbert5125gmail** [reply]
Java Learning Resources:
- Code academy
- Practice it

**Reply by ryanelbert5125gmail** [reply]
+ Wpilib java resources

**Reply by ryanelbert5125gmail** [reply]
- Make sure to ask frequently if people don't understand stuff so we can help them understand
- Make sure one working drive base
- Todo: get Krakens working
